design-proposals/crd-based-management.md

```diff
@@ -0,0 +1,281 @@
+
+### Export local service user story
+
+User wants to export a single-port local service "foo" in namespace X.
+
+To do so, it creates an Export k8s object whose name matches the name of the exported service (i.e. "foo").
+
+The status of the export can reflect the following errors:
+- The target service does not exists.
+- The target service has more/less than a single port.
+
+A pre-condition for the export to be consumed is that the actual service port is reported as part of the export status. 
+
+If any changes are made to the target service, the export status is expected to update accordingly.
+
+### Export multi-port local service user story
+
+User wants to export one or more ports from a multi-port local service "foo" in namespace X.
+
+To do so, it needs to create an Export k8s object for each port to be exported.
+The export object spec will include the name of the service ("foo") and the desired exported port.
+In case the service name is not specified in the spec, it is assumed to match the export name.
+
+The status of the export can reflect the following errors:
+- The target service does not exists.
+- The requested port is not a valid port in the requested service.
+
+Like in the single-port case above, the port (copied from the spec) is reported as part of the export status.
+If any changes are made to the target service, the export status is expected to update accordingly.
+
+### Export endpoint user story
+
+User wants to export some service which is accessible from within a cluster via an endpoint host/ip:port.
+In particular, this can be used to export a service from any namespace (e.g. my-service.my-namespace.svc.cluster.local).
+
+To do so, the user needs to create an Export k8s object whose spec includes the service host/ip and port.
+
+In this case, the export status is empty, as no validation is made against the endpoint.
+
+**QUESTION**: do we want to try to DNS lookup the service host, and post a failure in the status?
+
+### Peer creation user story
+
+Site A admin wants to allow users to import services from site B.
+
+To do so, it creates a Peer k8s object for site B.
+The name of the object will be used by importers as part of the routable name of an export.
+The peer spec includes one or more gateways (host/ip,port) for accessing the peer's dataplane server.
+
+**QUESTION**: Should we add an expected server name to the spec, or rely just on the gateway hosts?
+
+The status of the peer object will report whether the peer is currently reachable,
+and the last time it was successfully reached (via heartbeats).
+
+### Import service user story
+
+User of site B wants to import an exported service "foo" from namespace X of site A,
+into namespace Y of site B.
+
+To do so, it creates an Import k8s object in the desired namespace Y.
+The spec of the import includes one or more exports which serve as the source of the import.
+Each export is identified using its routable name: (peer name, export namespace, export name).
```

Comment:
What happens:
1. If he wants to export from multiple peers? need 2 imports?
2. If he doesn't provide the namespace?
3. If we have multiple ports? Do we open a connection per port?